polynote-frontend/polynote/cell.js

```diff
@@ -407,6 +423,52 @@ export class CodeCell extends Cell {
     get content() {
         return this.editor.getValue();
     }
+
+    applyEdits(edits) {
+        // can't listen to these edits or they'll be sent to the server again
+        // TODO: is there a better way to silently apply these edits? This seems like a hack; only works because of
+        //       single-threaded JS, which I don't know whether workers impact that assumption (JS)
+        this.applyingServerEdits = true;
+
+        try {
+            const monacoEdits = [];
+            const model = this.editor.getModel();
+            edits.forEach((edit) => {
+                if (edit.isEmpty()) {
+                    return;
+                }
+
+                const pos = model.getPositionAt(edit.pos);
+                if (edit instanceof messages.Delete) {
+                    const endPos = model.getPositionAt(edit.pos + edit.length);
+                    monacoEdits.push({
+                        range: new monaco.Range(pos.lineNumber, pos.column, endPos.lineNumber, endPos.column),
+                        text: null
+                    });
+                } else if (edit instanceof messages.Insert) {
+                    monacoEdits.push({
+                        range: new monaco.Range(pos.lineNumber, pos.column, pos.lineNumber, pos.column),
+                        text: edit.content,
+                        forceMoveMarkers: true
+                    });
```

Comment:
Hmm, instead of translating `Messages` to Monaco edit operations here, I wonder whether it might be nicer (from a maintainability perspective) to have some sort of `toMonacoEdit(model)` function on each message type. That way we can look at messages.js and see how messages get translated right next to their definitions?

Reply:
Good idea; I'll clean that up in another PR.

Reply:
(it's hardly the biggest thing that needs to be moved to a more sane place... see basically all of `ui.js` 😬 )

Reply:
I created https://github.com/polynote/polynote/issues/36 to keep track of UI issues.